spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/pubsub/PubSubBeanProcessor.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2021-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spring.autoconfigure.pubsub;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import com.google.cloud.spring.autoconfigure.core.GcpProperties;
+import com.google.cloud.spring.core.DefaultGcpProjectIdProvider;
+import com.google.cloud.spring.core.GcpProjectIdProvider;
+import com.google.cloud.spring.pubsub.core.PubSubConfiguration;
+import com.google.cloud.spring.pubsub.support.PubSubSubscriptionUtils;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.beans.factory.support.BeanDefinitionBuilder;
+import org.springframework.beans.factory.support.BeanDefinitionRegistry;
+import org.springframework.beans.factory.support.BeanDefinitionRegistryPostProcessor;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.boot.context.properties.bind.Binder;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
+
+/**
+ * Class to register ThreadPoolTaskScheduler beans.
+ */
+public class PubSubBeanProcessor implements BeanDefinitionRegistryPostProcessor {
+
+	private ConfigurableEnvironment environment;
+
+	private ConcurrentHashMap<String, ThreadPoolTaskScheduler> threadPoolTaskSchedulerMap = new ConcurrentHashMap<>();
+
+	private ThreadPoolTaskScheduler globalScheduler;
+
+	public PubSubBeanProcessor(ConfigurableEnvironment environment) {
+		this.environment = environment;
+	}
+
+	@Override
+	public void postProcessBeanDefinitionRegistry(BeanDefinitionRegistry beanDefinitionRegistry) throws BeansException {
+		Binder binder = Binder.get(this.environment);
+		String cloudPropertiesPrefix = GcpProperties.class.getAnnotation(ConfigurationProperties.class)
+				.value();
+		GcpProperties gcpProperties = binder.bind(cloudPropertiesPrefix, GcpProperties.class)
+				.orElse(new GcpProperties());
+		GcpProjectIdProvider projectIdProvider = (gcpProperties.getProjectId() != null)
```

Comment:
Slight wrinkle: This should actually be coming from  `spring.cloud.gcp.pubsub.project-id`, *or* [the default provider](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/62389625dc205ce3e431967249877c9c0f1fa600/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/core/GcpContextAutoConfiguration.java#L73).

Take another look at [pubsub autoconfiguration](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/0def25d2cac17a5045c0f8d4c4dd9d15059622aa/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/pubsub/GcpPubSubAutoConfiguration.java#L102) -- it's confusing, but the order of precedence is:
* product-specific property `spring.cloud.gcp.pubsub.project-id`
* global property `spring.cloud.gcp.project-id`
* default from environment

I'd separate the logic for determining project ID into a helper method here, comment in javadoc why the duplication is needed, *and* possibly even add a few unit tests for autoconfiguration verifying that this project ID logic will get the same result as the regular autoconfiguration. The reason for the unit tests is that we'll need them to fail to signal that the two project ID sets of logic have diverged. Otherwise we'll never remember to keep them in sync because they Almost Never Change.

Reply:
Is there any way to have this post processor run after Spring has already processed `GcpPubSubAutoConfiguration`? Maybe you can create another auto-configuration class for the post-processor where you do `@AutoConfigureAfter(GcpPubSubAutoConfiguration.class)`? I'm not sure if it would work, but it would nice not have to duplicate this logic.

Reply:
> Slight wrinkle: This should actually be coming from `spring.cloud.gcp.pubsub.project-id`, _or_ [the default provider](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/62389625dc205ce3e431967249877c9c0f1fa600/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/core/GcpContextAutoConfiguration.java#L73).

Ah you're right. There is a lot more logic involved here which unfortunately may have to be duplicated.

> Is there any way to have this post processor run after Spring has already processed `GcpPubSubAutoConfiguration`

That's an interesting thought! Unfortunately, `@AutoConfigureAfter(GcpPubSubAutoConfiguration.class)` still results in the BeanDefinitionRegistryPostProcessor occurring before any of the configurations. The order looks like: BDRPP ->  GcpContextAutoConfiguration --> GcpPubSubAutoConfiguration --> "Configuration annotated with AfterConfigureAfter"